File: tobas_motor_test/src/tobas_motor_test/rotor_speeds_publisher.py
import rospy
from typing import List
from functools import partial
from PyQt5.QtCore import pyqtSlot
from PyQt5.QtWidgets import QPushButton, QVBoxLayout

from tobas_std_tools_py.math import rps2rpm, rpm2rps
from tobas_tools_py.constants import Topic
from tobas_tools_py.drone import Drone, DroneLoader_Param
from tobas_rqt_tools.widgets import Widget, IntSliderDisplay
from tobas_msgs.msg import RotorSpeeds

from .common import BUTTON_HEIGHT, COMMAND_PERIOD


class RotorSpeedsPublisherWidget(Widget):
    def __init__(self) -> None:
        super().__init__()

        drone = Drone()
        DroneLoader_Param(drone).load()

        rows = QVBoxLayout()
        self.setLayout(rows)

        self._commanders: List[IntSliderDisplay] = []
        for rotor in drone.rotors:
            commander = IntSliderDisplay(self)
            commander.set_text(f"CH{rotor.channel}")
            commander.set_minimum(0)
            commander.set_maximum(rps2rpm(rotor.max_rot_speed))
            commander.set_value(0)
            commander.set_suffix(" rpm")
            commander.value_changed.connect(self._on_value_changed)
            rows.addWidget(commander)
            self._commanders.append(commander)

        for rpm in [0, 100, 500, 1000, 5000, 10000]:
            button = QPushButton(f"{rpm} RPM")
            button.setFixedHeight(BUTTON_HEIGHT)
            button.clicked.connect(partial(self._on_rpm_button_clicked, rpm=rpm))
            rows.addWidget(button)

        rows.addStretch()

        self._speeds_pub = rospy.Publisher(Topic.Command.ROTOR_SPEEDS, RotorSpeeds, queue_size=1)

        # モータが停止しないよう一定周期でコマンドを発行し続ける
        rospy.Timer(rospy.Duration(COMMAND_PERIOD), self._command_timer_cb)

    @pyqtSlot()
    def _on_value_changed(self) -> None:
        self._publish_current_values()

    @pyqtSlot()
    def _on_rpm_button_clicked(self, rpm: int) -> None:
        self._set_all_values(rpm)
        self._publish_current_values()

    def _publish_current_values(self) -> None:
        rot_speeds = RotorSpeeds()
        rot_speeds.header.stamp = rospy.Time.now()
        rot_speeds.speeds = [0.0] * len(self._commanders)

        for channel, commander in enumerate(self._commanders):
            rot_speeds.speeds[channel] = rpm2rps(commander.get_value())

        self._speeds_pub.publish(rot_speeds)

    def _set_all_values(self, value: int) -> None:
        for commander in self._commanders:
            commander.blockSignals(True)
            commander.set_value(value)
            commander.blockSignals(False)

    def _command_timer_cb(self, _) -> None:
        self._publish_current_values()

File: tobas_gui/tobas_hardware_setup/src/rotor_test/rotor_speeds_publisher.cpp
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGridLayout>

#include <tobas_std_tools/unit_conversions.hpp>
#include <tobas_path_tools/join.hpp>

#include "tobas_hardware_setup/rotor_test/rotor_speeds_publisher.hpp"

namespace gui
{
namespace hardware_setup
{
RotorSpeedsPublisherWidget::RotorSpeedsPublisherWidget(rclcpp::Node::SharedPtr node, const tobas::Drone& drone)
  : node_(node), drone_(drone)
{
  const auto rows = new QVBoxLayout();
  setLayout(rows);

  const auto grid = new QGridLayout();
  rows->addLayout(grid);

  for (size_t ch = 0; ch < kChannelSize; ++ch)
  {
    commanders_[ch] = new qt::IntSliderDisplay();
    commanders_[ch]->setSuffix(" rpm");
    connect(commanders_[ch], &qt::IntSliderDisplay::valueChanged, this, &self::onValueChanged);
    grid->addWidget(commanders_[ch], ch % kMaxRows, ch / kMaxRows);
  }

  const auto cols = new QHBoxLayout();
  rows->addLayout(cols);

  for (size_t i = 0; i < kNumButtons; ++i)
  {
    const auto& rpm = kRPMs[i];
    rpm_buttons_[i] = new QPushButton(QString::number(rpm) + " RPM");
    rpm_buttons_[i]->setFixedHeight(kButtonHeight);
    connect(rpm_buttons_[i], &QPushButton::clicked, bind(&self::onRPMButtonClicked, this, rpm));
    cols->addWidget(rpm_buttons_[i]);
  }
}

void RotorSpeedsPublisherWidget::updateInternalDataStructures()
{
  // モータとして登録されているチャンネルの設定
  std::unordered_set<size_t> rotor_channels;
  for (const auto& rotor : drone_.rotors)
  {
    const auto& ch = rotor.channel;
    rotor_channels.insert(ch);
    commanders_.at(ch)->setText("CH" + QString::number(ch) + ": " + QString::fromStdString(rotor.link_name));
    commanders_.at(ch)->setMaximum(tobas_std::rps2rpm(rotor.max_rot_speed));
    commanders_.at(ch)->setValue(0);
  }

  // モータとして登録されていないチャンネルを無効化
  for (size_t ch = 0; ch < kChannelSize; ++ch)
  {
    if (rotor_channels.contains(ch))
      continue;
    commanders_.at(ch)->setText("CH" + QString::number(ch) + ": unregistered");
    commanders_.at(ch)->setMaximum(0);
    commanders_.at(ch)->setEnabled(false);
  }

  // 回転数トピックを更新
  const auto speeds_topic = path::join(drone_.name, tobas::kRotorSpeedsCmdTopic);
  speeds_pub_ = ros2::createPublisher<tobas_msgs::msg::RotorSpeeds>(node_, speeds_topic);
}

void RotorSpeedsPublisherWidget::start()
{
  // コマンダーを有効化
  for (const auto& rotor : drone_.rotors)
  {
    commanders_.at(rotor.channel)->setValue(0);
    commanders_.at(rotor.channel)->setEnabled(true);
  }

  // RPMボタンを有効化
  for (const auto& button : rpm_buttons_)
    button->setEnabled(true);

  // モータが停止しないよう一定周期でコマンドを発行し続ける
  publish_timer_ = ros2::createTimer(node_, kPublishPeriod, &self::publishTimerCb, this);
}

void RotorSpeedsPublisherWidget::stop()
{
  // コマンダーを無効化
  for (const auto& rotor : drone_.rotors)
  {
    commanders_.at(rotor.channel)->setValue(0);
    commanders_.at(rotor.channel)->setEnabled(false);
  }

  // RPMボタンを無効化
  for (const auto& button : rpm_buttons_)
    button->setEnabled(false);

  // タイマーを停止
  publish_timer_->cancel();
}

void RotorSpeedsPublisherWidget::setAllValues(int value)
{
  for (const auto& rotor : drone_.rotors)
    commanders_.at(rotor.channel)->setValue(value);
}

void RotorSpeedsPublisherWidget::publishCurrentValues()
{
  auto rot_speeds = std::make_unique<tobas_msgs::msg::RotorSpeeds>();
  rot_speeds->header.stamp = node_->get_clock()->now();
  rot_speeds->speeds.resize(drone_.numRotors());

  for (const auto& rotor : drone_.rotors)
    rot_speeds->speeds.at(rotor.channel) = tobas_std::rpm2rps(commanders_.at(rotor.channel)->getValue());

  speeds_pub_->publish(std::move(rot_speeds));
}

void RotorSpeedsPublisherWidget::publishTimerCb()
{
  publishCurrentValues();
}

void RotorSpeedsPublisherWidget::onValueChanged()
{
  publishCurrentValues();
}

void RotorSpeedsPublisherWidget::onRPMButtonClicked(int rpm)
{
  setAllValues(rpm);
  publishCurrentValues();
}
}  // namespace hardware_setup
}  // namespace gui
